### PR TITLE
sonarr: build from source

### PR DESCRIPTION
**Pull request by tie**
## Description of changes

This change refactors `sonarr` package (and its update script) to build the package from source. It also contains changes to the underlying .NET build infrastructure required for this package.

See also https://github.com/NixOS/nixpkgs/pull/278050#issuecomment-1880438199

>[!IMPORTANT]
>This change relies on the recent PRs for the .NET build infrastructure (for structured attributes and cross-compilation) and cannot be backported.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by bjornfor** [comment]
> buildDotnetModule: do not run dotnet command using env 

Why not? Why was using `env` added in the first place? (Might be a good idea to answer that in the commit message.)

**Comment by tie** [comment]
W.r.t .NET build infrastructure updates. Some packages failed to build their web UI likely because of #253156, but otherwise looks good.

---

Result of `nixpkgs-review pr 291640` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>audiobookshelf</li>
    <li>jellyfin</li>
    <li>kavita</li>
    <li>slskd</li>
    <li>sonarr</li>
  </ul>
</details>
<details>
  <summary>60 packages built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
    <li>BeatSaberModManager</li>
    <li>alttpr-opentracker</li>
    <li>avalonia-ilspy</li>
    <li>azure-functions-core-tools</li>
    <li>boogie (dotnetPackages.Boogie)</li>
    <li>btcpayserver</li>
    <li>btcpayserver-altcoins</li>
    <li>cavalier</li>
    <li>celeste64</li>
    <li>certdump</li>
    <li>csharp-ls</li>
    <li>csharpier</li>
    <li>csharprepl</li>
    <li>dafny</li>
    <li>denaro</li>
    <li>depotdownloader</li>
    <li>discordchatexporter-cli</li>
    <li>eventstore</li>
    <li>fable</li>
    <li>fantomas</li>
    <li>formula</li>
    <li>fsautocomplete</li>
    <li>galaxy-buds-client</li>
    <li>git-credential-manager</li>
    <li>github-runner</li>
    <li>ilspycmd</li>
    <li>inklecate</li>
    <li>jackett</li>
    <li>knossosnet</li>
    <li>libation</li>
    <li>lubelogger</li>
    <li>marksman</li>
    <li>mqttmultimeter</li>
    <li>msbuild</li>
    <li>naps2</li>
    <li>nbxplorer</li>
    <li>netcoredbg</li>
    <li>networkminer</li>
    <li>nuget-to-nix</li>
    <li>omnisharp-roslyn</li>
    <li>openra</li>
    <li>opentabletdriver</li>
    <li>openutau</li>
    <li>osu-lazer</li>
    <li>parabolic</li>
    <li>pbm</li>
    <li>pinta</li>
    <li>ps3-disc-dumper</li>
    <li>pupdate</li>
    <li>roslyn</li>
    <li>roslyn-ls</li>
    <li>ryujinx</li>
    <li>scarab</li>
    <li>space-station-14-launcher</li>
    <li>tests.dotnet.project-references</li>
    <li>tone</li>
    <li>torrentstream</li>
    <li>wasabibackend</li>
    <li>xivlauncher</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Is there a particular reason you are targeting staging here?

**Comment by tie** [comment]
I was expecting more rebuilds from .NET changes 😅

**Comment by purcell** [comment]
Not sure I can meaningfully review this, but as one of the `sonarr` maintainers this is a great step, and the specific changes to the Sonarr derivation look good to me.

File: pkgs/by-name/so/sonarr/package.nix
Comment (purcell) [inline pkgs/by-name/so/sonarr/package.nix:85]: In the previous version the executable was (rather unfortunately) called `NzbDrone`, and the systemd unit refers to that too IIRC, so you might need to update that too.

Comment (tie) [reply pkgs/by-name/so/sonarr/package.nix:85]: Done. Though we also create NzbDrone symlink in postFixup hook for compatibility.

**Comment by purcell** [comment]
I tried running this on MacOS but can't execute the resulting `Sonarr` executable, strangely:
```
❯ nix run 'github:tie/nixpkgs/sonarr-v4#sonarr'
zsh: killed     nix run 'github:tie/nixpkgs/sonarr-v4#sonarr'
```
or, more directly,
```
❯ /nix/store/m7ykc6igjqq9ivnb6rca0qlz8njr68fp-sonarr-4.0.1.929/bin/Sonarr
zsh: killed     /nix/store/m7ykc6igjqq9ivnb6rca0qlz8njr68fp-sonarr-4.0.1.929/bin/Sonarr
```
and there's a Console message saying
```
ASP: Security policy would not allow process: 65161, /nix/store/m7ykc6igjqq9ivnb6rca0qlz8njr68fp-sonarr-4.0.1.929/lib/sonarr/Sonarr
```
In contrast, this correctly starts the server:
```
❯ nix run 'github:NixOS/nixpkgs/nixpkgs-unstable#sonarr'
```
Perhaps there's a codesigning step missing for darwin, but I'mn not an expert on such things sorry.

**Comment by tie** [comment]
@purcell, weird, I can’t reproduce this in macOS VM (but then I guess it may have relaxed security constraints). Do other .NET packages work for you (e.g. jellyfin)?

**Comment by purcell** [comment]
> weird, I can’t reproduce this in macOS VM (but then I guess it may have relaxed security constraints).

I have SIP enabled, which would be the MacOS default, and I have the sandbox enabled too.

> Do other .NET packages work for you (e.g. jellyfin)?

Yes, `nix run 'github:NixOS/nixpkgs#jellyfin'` executes fine.

**Comment by tie** [comment]
@purcell, looks like something related to .NET is actually broken in upstream Nixpkgs, not sure why binary cache is fine though.

<details>
<summary>System information</summary>

```
; git rev-parse HEAD
bd5ddf2c6bfafff031edf80221e1ee94e86ca10a
; git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
; system_profiler SPSoftwareDataType
Software:

    System Software Overview:

      System Version: macOS 14.2.1 (23C71)
      Kernel Version: Darwin 23.2.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: ryu
      User Name: Ivan Trubach (tie)
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
```
</details>

```
; nix build --verbose --file . jellyfin
this path will be fetched (10.93 MiB download, 43.05 MiB unpacked):
  /nix/store/znamd3abcaszf0p8xdin9g8y16jk1nsg-jellyfin-10.8.13
copying path '/nix/store/znamd3abcaszf0p8xdin9g8y16jk1nsg-jellyfin-10.8.13' from 'https://cache.nixos.org'...
; ./result/bin/jellyfin --help
```
Works fine, jellyfin is fetched from binary cache.

But then
```
; nix build --file . --rebuild jellyfin
error: derivation '/nix/store/1ljiqchsj7flkgzsc0063n54nb40av9k-jellyfin-10.8.13.drv' may not be deterministic: output '/nix/store/znamd3abcaszf0p8xdin9g8y16jk1nsg-jellyfin-10.8.13' differs
```
and to confirm:
```
; nix store delete /nix/store/znamd3abcaszf0p8xdin9g8y16jk1nsg-jellyfin-10.8.13
1 store paths deleted, 43.02 MiB freed
; nix build --file . --offline jellyfin
; ./result/bin/jellyfin 
zsh: killed     ./result/bin/jellyfin
```

**Comment by purcell** [comment]
Ugh, that's unfortunate. But good investigation skills! Any idea what we could do to escalate this or find out more?

**Comment by purcell** [comment]
I would think that if the official darwin builders don't enable the sandbox, then as long as this PR works with the sandbox disabled, it shouldn't block us from potentially merging. So I'd suggest we resolve the conflicts that have arisen here, and then find a merger.

**Comment by tie** [comment]
Resolved conflicts and rebased.

File: nixos/modules/services/misc/sonarr.nix
Comment (ambroisie) [inline nixos/modules/services/misc/sonarr.nix:57]: I think the _technically correct_ way to do this is `utils.escapeSystemdExecArgs` (`utils` being a NixOS `specialArg`, like `lib`, `config`, or `pkgs`)

Comment (tie) [reply nixos/modules/services/misc/sonarr.nix:57]: Oh, thank you, I didn’t known this function exists!
```
$ cd nixpkgs/nixos
$ rg -l escapeShellArgs | wc -l
123
$ rg -l escapeSystemdExecArgs | wc -l
20
```
Doesn’t seem like it’s used in nixpkgs as much as it should be, unfortunately.

**Comment by tie** [comment]
I’ve requested a review from @corngood, author of the [last non-trivial PR for .NET build infrastructure](https://github.com/NixOS/nixpkgs/pull/190129), since ideally I’d like to see a review from someone familiar with this stuff.

@corngood, perhaps you have some ideas on https://github.com/NixOS/nixpkgs/pull/291640#issuecomment-1997181117

**Comment by corngood** [comment]
I have #290684 which may fix dotnet the sandbox problems.  It looks like you're getting an executable with a broken signature.

File: pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix
Comment (corngood) [inline pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix:24]: I don't think we need this comment.  It's fine if you describe this in the PR.

File: pkgs/by-name/so/sonarr/package.nix
Comment (corngood) [inline pkgs/by-name/so/sonarr/package.nix:73]: I think instead of adding support for `nugetConfigFile`, it could be a lot simpler to either:

- fix whatever is not finding the config file in src/

or 

- copy/link src/NuGet.Config into the root 

My worry is that any use of dotnet tools that isn't aware of this flag will be broken.

**Review by corngood — CHANGES_REQUESTED** [review]

File: pkgs/by-name/so/sonarr/buildprops.patch
Comment (GGG-KILLER) [inline pkgs/by-name/so/sonarr/buildprops.patch:1]: Instead of a patch, these could be done with `-p:PropName=Value` args, as that overrides any props set in config.

Don't know if it'd be easier to do that or maintain a patch.

File: pkgs/by-name/so/sonarr/package.nix
Comment (GGG-KILLER) [reply pkgs/by-name/so/sonarr/package.nix:73]: I agree with this, moving the file to the root would be much better than changing the entire dotnet infra to accomodate a single pacakge.

Comment (tie) [reply pkgs/by-name/so/sonarr/package.nix:73]: We need this for `nuget-to-nix` that uses `dotnet nuget list source` and therefore does not search for NuGet configuration along the project file path. Perhaps we can teach `nuget-to-nix` to do that?

File: pkgs/by-name/so/sonarr/buildprops.patch
Comment (tie) [reply pkgs/by-name/so/sonarr/buildprops.patch:1]: Done.

**Comment by tie** [comment]
I’ve converted the PR to draft since I had to fix some other packages after rebasing and I’d like to submit PRs for them separately. Also, buildDotnetModule changes kinda gone out of the control with all the changes to support structured attributes for `Copyright` property.

**Comment by tie** [comment]
Note: actually fix for dbus-python cross-compilation is already merged in #308196. I’ll try to move as much .NET changes as possible to separate PRs.

File: pkgs/by-name/so/sonarr/package.nix
Comment (tie) [reply pkgs/by-name/so/sonarr/package.nix]: Added a patch to move NuGet.config to the source root.

**Comment by tie** [comment]
Result of `nixpkgs-review pr 291640` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>sonarr</li>
  </ul>
</details>

**Comment by tie** [comment]
@purcell, this should be ready for review and without .NET infrastructure changes. Note that these were merged separately (huge thanks to .NET build infrastructure maintainers for fast review cycles) and are required for this PR. That is, Sonarr version updates for the release channel would have to be done separately (see also note in PR description).

**Comment by purcell** [comment]
Sounds great to me, and I very much support this, but I'm a bit too slammed currently to spin it up locally for testing.

File: pkgs/by-name/so/sonarr/package.nix
Comment (purcell) [inline pkgs/by-name/so/sonarr/package.nix:157]: Thank goodness you're adding yourself as a maintainer for this, because maintaining this new version of the derivation is likely beyond my wizard level.

**Review by corngood — APPROVED** [review]

File: pkgs/by-name/so/sonarr/package.nix
Comment (tie) [reply pkgs/by-name/so/sonarr/package.nix:157]: Yes, I’m, unfortunately, stuck with Sonarr 😅
It’s sad that it’s not fully self-hostable (it requires skyhook.sonarr.tv to be up and reachable for operation), but there are no other alternatives with comparable feature set, so it’s an important piece of infrastructure for me.

I don’t think there would be any non-trivial maintenance with source build aside from occasionally bumping .NET version, but feel free to ping me if there are any issues (besides, ofborg will ping me anyway).

**Comment by corngood** [comment]
> Sounds great to me, and I very much support this, but I'm a bit too slammed currently to spin it up locally for testing.

@purcell I'd like your approval before merging.  Should we wait for you to test it?

File: pkgs/by-name/so/sonarr/package.nix
Comment (corngood) [reply pkgs/by-name/so/sonarr/package.nix:157]: I'm happy to help with general dotnet source-build stuff too.

**Review by purcell — APPROVED** [review]
I built and ran this without issues on `aarch64-darwin`, and given the state of the CI for this PR, I'm going to approve it and suggest we merge.

**Comment by purcell** [comment]
Thanks @tie! (and @corngood for the merge)

**Pull request merged**
